File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationLibraryLoader.java
package com.mapbox.services.android.navigation.v5.navigation;

import timber.log.Timber;

public abstract class NavigationLibraryLoader {

  private static final String NAVIGATION_NATIVE = "navigator-android";
  private static final NavigationLibraryLoader DEFAULT = new NavigationLibraryLoader() {
    @Override
    public void load(String name) {
      System.loadLibrary(name);
    }
  };

  private static volatile NavigationLibraryLoader loader = DEFAULT;

  /**
   * Loads navigation shared library.
   * <p>
   * Catches UnsatisfiedLinkErrors and prints a warning to logcat.
   * </p>
   */
  public static void load() {
    try {
      loader.load(NAVIGATION_NATIVE);
    } catch (UnsatisfiedLinkError error) {
      Timber.e(error, "Failed to load native shared library.");
    }
  }

  public abstract void load(String name);
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationLibraryLoader.kt
package com.mapbox.services.android.navigation.v5.navigation

import com.mapbox.navigation.base.logger.model.Message
import com.mapbox.navigation.logger.MapboxLogger

abstract class NavigationLibraryLoader {

    abstract fun load(name: String)

    companion object {

        private val NAVIGATION_NATIVE = "navigator-android"
        private val DEFAULT: NavigationLibraryLoader = object : NavigationLibraryLoader() {
            override fun load(name: String) {
                System.loadLibrary(name)
            }
        }

        @Volatile
        private var loader = DEFAULT

        /**
         * Loads navigation shared library.
         *
         *
         * Catches UnsatisfiedLinkErrors and prints a warning to logcat.
         *
         */
        fun load() {
            try {
                loader.load(NAVIGATION_NATIVE)
            } catch (error: UnsatisfiedLinkError) {
                MapboxLogger.e(Message("Failed to load native shared library."), error)
            }
        }
    }
}
